src/test/ui/if/if-without-else-as-fn-expr.stderr

```diff
@@ -0,0 +1,17 @@
+error[E0317]: if may be missing an else clause
+  --> $DIR/if-without-else-as-fn-expr.rs:2:5
+   |
+LL |   fn foo(bar: usize) -> usize {
+   |                         ----- found `usize` because of this return type
+LL | /     if bar % 5 == 0 {
+LL | |         return 3;
+LL | |     }
+   | |_____^ expected (), found usize
+   |
+   = note: expected type `()`
+              found type `usize`
+   = note: `if` expressions without `else` must evaluate to `()`
```

Comment:
I found this error message was still quite confusing initially. It's clear to me that the function will return `()` and not `usize` when the `if` condition isn't `true` and that this is the error, but perhaps outwith the `if may be missing an else clause` part of the error, this isn't well articulated. I think there are two reasons for this:

- It appears to me that we should expect a `usize` (due to the return type) and have found a `()` (due to the missing `else` block). Currently the error suggests the reverse, which is correct when we only consider what would be required to have the `if` without an `else` be valid.

- While the note is factually correct, I initially interpreted it as a suggestion rather than a statement of fact - that if I were to change the `if` block to evaluate to `()` then this would fix the error, but that isn't the case. A wording such as: `` `if` expressions can only omit an `else` block when they are expected to evaluate to `()` `` would clear that up.

I think something along these lines could be more helpful for this particular test:

```
error[E0317]: if may be missing an else clause
  --> $DIR/if-without-else-as-fn-expr.rs:2:5
   |
LL |   fn foo(bar: usize) -> usize {
   |                         ----- expected `usize` because of this return type
LL | /     if bar % 5 == 0 {
LL | |         return 3;
LL | |     }
   | |_____^ expected usize, found ()
   |
   = note: expected type `usize`
              found type `()`
   = note: this function evaluates to `()` when the `if` condition is not `true`...
   = note: ...consider adding a `else` block that evaluates to `usize`.
```

More generally, I think when an `if` expression is expected to evaluate to `()` then the note as in this PR currently is excellent, but if it doesn't match, we should instead explain that a implicit `else { }` block is returning the wrong type, and an `else` block should be added.

Reply:
@davidtwco let me know what you think of [the new output](https://github.com/rust-lang/rust/pull/58981/files#diff-13bacecdd7126491beeb136d2ce6532b). It's not exactly the same you requested, as I was trying to avoid incorrect diagnostics in some cases where the coercion is failing from within the arm, not outside.

Reply:
Thanks, that's perfect.